<a href="https://colab.research.google.com/github/dazzens/dazzens/blob/main/EDA_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

------------------
---------
-----

# **Пет проект EDA (исследовательский анализ данных).**




Автор: А.Г.Х.

Начало: 22.06.2026


Окончание: 24.06.2026

----



##**Бизнес-контекст:**

- Интернет-магазин продает товары для дома и декора по всему миру. За последний год ассортимент вырос на 30%, но темпы роста выручки замедлились. Необходимо исследовать данные о продажах, чтобы понять, какие товары и страны приносят основную прибыль, кто самые ценные клиенты и как оптимизировать ценовую политику.

-----

# **Цель исследования**

- Выявить ключевые факторы, влияющие на динамику продаж интернет-магазина, для разработки рекомендаций по увеличению выручки и удержанию клиентов.


##**Вопросы:**


- 1. Как менялись продажи по месяцам? Динамика выручки, количества заказов и сезонность.


- 2. Какие товары самые популярные? Топ по количеству продаж, топ по выручке, топ по частоте заказов.

- 3. Какой средний чек и размер корзины? Среднее количество товаров в заказе, средняя сумма заказа

- 4. Какие страны приносят больше всего выручки? Топ по выручке, топ по количеству заказов, топ по среднему чеку

- 5. Кто самые прибыльные клиенты? Топ клиентов по сумме трат, по частоте покупок, анализ повторных покупок

- 6. Как цена влияет на продажи? Распределение цен, зависимость количества продаж от цены, оптимальные ценовые сегменты

- 7. Какие товары, страны и клиентские сегменты вносят наибольший вклад в выручку?


## **Содержание проекта**



1.  **Первичный осмотр и подготовка данных**
    *   Импорт библиотек и загрузка данных.
    *   Предварительный анализ структуры данных (типы, объем, пропуски).
    *   Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
    *   Проверка и обработка пропусков (`CustomerNo`).
    *   Проверка дубликатов и выбросов (`Price`, `Quantity`).

2.  **Предварительный анализ и создание новых признаков**
    *   Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
    *   Создание признаков: месяц, день недели, год из даты.
    *   Создание признака: общая сумма заказа (`Price` x `Quantity`).
    *   Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

3.  **Исследовательский анализ данных (EDA)**
    *   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
    *   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
    *   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
    *   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
    *   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
    *   Анализ ценовых сегментов: зависимость количества продаж от цены.

4.  **Ключевые выводы и рекомендации**
    *   Формирование основных инсайтов по каждому блоку анализа.
    *   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.


    

## **Описание данных:**

**Транзакционные данные (онлайн-магазин):**

| | | |
|------|----------|------------|
| `TransactionNo` | Уникальный идентификатор транзакции/заказа | `object` |
| `Date` | Дата совершения покупки | `object` |
| `ProductNo` | Уникальный идентификатор товара | `object` |
| `ProductName` | Название товара | `object` |
| `Price` | Цена за единицу товара | `float64` |
| `Quantity` | Количество единиц товара в заказе | `int64` |
| `CustomerNo` | Уникальный идентификатор клиента | `float64` |
| `Country` | Страна клиента | `object` |

---

**Характеристики данных:**

|  |  |
|------------|----------|
| **Объем данных** | 536 350 записей |
| **Количество клиентов** | 4 738 |
| **Количество уникальных товаров** | 3 768 |
| **Количество стран** | 38 |
| **Пропуски** | 55 записей без `CustomerNo` (0.01%) |



----
----

# **Шаг 1. Первичный осмотр и подготовка данных**

- Импорт библиотек и загрузка данных.
- Предварительный анализ структуры данных (типы, объем, пропуски).
- Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
- Проверка и обработка пропусков (`CustomerNo`).
- Проверка дубликатов и выбросов (`Price`, `Quantity`).


---
## **Библиотеки**




In [1]:

# Необходимые библиотеки

try:
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    from scipy.stats import ttest_ind
    from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Все библиотеки загружены")

except ImportError:
    !pip install numpy pandas matplotlib seaborn scipy matplotlib-venn kagglehub -q
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    from scipy.stats import ttest_ind
    from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Библиотеки загружены через pip")



Все библиотеки загружены


# // Оптимизировать загрузку библиотек ^^^^^

## **Загрузка датасета:**

In [2]:
# ============================================
# ШАГ 2: Скачиваем датасет через kagglehub
# ============================================
print("⏳ Скачивание датасета...")
path = kagglehub.dataset_download("gabrielramos87/an-online-shop-business")
print(f"✅ Путь к данным: {path}")

⏳ Скачивание датасета...


100%|██████████| 6.66M/6.66M [00:00<00:00, 9.00MB/s]

Extracting files...


✅ Путь к данным: /root/.cache/kagglehub/datasets/gabrielramos87/an-online-shop-business/versions/7


In [3]:

# ============================================
# ШАГ 3: Смотрим все файлы в папке
# ============================================
print("\n📁 Найденные файлы:")
all_files = os.listdir(path)
for file in all_files:
    print(f"  - {file}")


📁 Найденные файлы:
  - Sales Transaction v.4a.csv


In [4]:
# ============================================
# ШАГ 4: Загружаем CSV файл
# ============================================
# Ищем все CSV файлы
csv_files = [f for f in all_files if f.endswith('.csv')]

if csv_files:
    # Берем первый CSV файл (это и есть ваш Sales Transaction v.4a.csv)
    file_name = csv_files[0]
    file_path = os.path.join(path, file_name)

    print(f"\n📂 Загружаем файл: {file_name}")
    df = pd.read_csv(file_path)


📂 Загружаем файл: Sales Transaction v.4a.csv


#// упростить загрузку ДАТАСЕТА ^^^^^

-----
----
----

## **Общий осмотр данных**

In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [48]:
df.head(10)



,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
0,581482,2019-12-09,22485,Set Of 2 Wooden Market Crates,21.47,12,17490,United Kingdom
1,581475,2019-12-09,22596,Christmas Star Wish List Chalkboard,10.65,36,13069,United Kingdom
2,581475,2019-12-09,23235,Storage Tin Vintage Leaf,11.53,12,13069,United Kingdom
3,581475,2019-12-09,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069,United Kingdom
4,581475,2019-12-09,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069,United Kingdom
5,581475,2019-12-09,21705,Bag 500g Swirly Marbles,10.65,24,13069,United Kingdom
6,581475,2019-12-09,22118,Joy Wooden Block Letters,11.53,18,13069,United Kingdom
7,581475,2019-12-09,22119,Peace Wooden Block Letters,12.25,12,13069,United Kingdom
8,581475,2019-12-09,22217,T-Light Holder Hanging Lace,10.65,12,13069,United Kingdom
9,581475,2019-12-09,22216,T-Light Holder White Lace,10.55,24,13069,United Kingdom


In [9]:
df.sample(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
477109,541423,1/17/2019,21200,Multicolour Honeycomb Paper Garland,11.92,4,13423.0,United Kingdom
303901,557628,6/21/2019,82600,N0 Singing Metal Sign,12.40,2,15311.0,United Kingdom
302160,557670,6/22/2019,22489,Pack Of 12 Traditional Crayons,10.62,432,12798.0,Japan
173261,568952,9/29/2019,22812,Pack 3 Boxes Christmas Panettone,12.25,24,14305.0,United Kingdom
171359,569202,9/30/2019,21928,Jumbo Bag Scandinavian Blue Paisley,7.24,1,12202.0,United Kingdom
335122,554347,5/24/2019,23199,Jumbo Bag Apples,6.19,30,16523.0,United Kingdom
343632,553679,5/18/2019,23284,Doormat Keep Calm And Come In,17.17,10,14565.0,United Kingdom
408292,547327,3/22/2019,22329,Round Container Set Of 5 Retrospot,11.94,12,12681.0,France
273968,560203,7/15/2019,22615,Pack Of 12 Circus Parade Tissues,10.55,2,15311.0,United Kingdom
431753,545333,3/1/2019,90175D,Tigris Eye Chunky Charm Bracelet,19.63,1,17333.0,United Kingdom


In [10]:
#предварительный осомтр
df.describe()


,Price,Quantity,CustomerNo
count,536350.000000,536350.000000,536295.000000
mean,12.662182,9.919347,15227.893178
std,8.490450,216.662300,1716.582932
min,5.130000,-80995.000000,12004.000000
25%,10.990000,1.000000,13807.000000
50%,11.940000,3.000000,15152.000000
75%,14.090000,10.000000,16729.000000
max,660.620000,80995.000000,18287.000000


In [41]:
display(df[df['Quantity'] < 0].head(15).sort_values(by='Quantity', ascending=False))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
1620,C581569,2019-12-09,84978,Hanging Heart Jar T-Light Holder,6.19,-1,17315,United Kingdom
1619,C581568,2019-12-09,21258,Victorian Sewing Box Large,6.19,-5,15311,United Kingdom
1621,C581569,2019-12-09,20979,36 Pencils Tube Red Retrospot,6.19,-5,17315,United Kingdom
6476,C581228,2019-12-08,22423,Regency Cakestand 3 Tier,6.19,-6,16019,United Kingdom
6478,C581228,2019-12-08,82494L,Wooden Frame Antique White,6.19,-6,16019,United Kingdom
1618,C581490,2019-12-09,23144,Zinc T-Light Holder Stars Small,6.04,-11,14397,United Kingdom
1617,C581490,2019-12-09,22178,Victorian Glass Hanging T-Light,6.19,-12,14397,United Kingdom
6477,C581228,2019-12-08,23210,White Rocking Horse Hand Painted,6.19,-12,16019,United Kingdom
6483,C581229,2019-12-08,23392,Spaceboy Rocket Lolly Makers,6.19,-24,12558,USA
6479,C581228,2019-12-08,22781,Gumball Magazine Rack,6.19,-24,16019,United Kingdom


In [46]:
# промежуточный анализ


sum_quantity_min = (df['Quantity'] < 0).sum()
count_quantity_min = (df['Quantity'] < 0).count()

print(sum_quantity_min)
print(count_quantity_min)

sum_negative = df[df['Quantity'] < 0]['Quantity'].sum()
print(f"Сумма отрицательных Quantity: {sum_negative}")


#######
######
#подсчет отрицательных колонок


negative_count = (df['Quantity'] < 0).sum()
print(f"Отрицательных Quantity: {negative_count:}")

# Доля отрицательных Quantity в процентах
negative_pct = (df['Quantity'] < 0).mean() * 100
print(f"Доля отрицательных Quantity: {negative_pct:.2f}%")

8585
536350
Сумма отрицательных Quantity: -271256
Отрицательных Quantity: 8585
Доля отрицательных Quantity: 1.60%


# **// Понять природу отрицательного числа количества**

// сделать предположение почему оно появилосб и дополнить краткий вывод на предосмотре

### **Осмотр уникальных позиций**

In [11]:
# Быстрый взгляд на все категориальные поля
df[['Country',
    'CustomerNo',
    'ProductName',
    'TransactionNo']].nunique().sort_values(ascending=False)


,0
TransactionNo,23204
CustomerNo,4738
ProductName,3768
Country,38


### **Анализ хвостов**

In [12]:
#Анализ хвоста цены

df['Price'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])

,Price
0.25,10.99
0.50,11.94
0.75,14.09
0.90,17.37
0.95,19.42
0.99,27.30


In [43]:

df['Quantity'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])

,Quantity
0.25,1.0
0.50,3.0
0.75,10.0
0.90,24.0
0.95,29.0
0.99,100.0


### **В результате предварительно осмотра установлено:**

- Объём выборки: 536 350 записей, 8 признаков.
- Пропуски: 0.01%  `CustomerNo`
- Размерность данных: 23 204 транзакции, 4 738 клиентов, 3 768 товаров, 38 стран.
- Распределение цен: медиана — 11.94 руб., 90-й перцентиль — 17.37 руб., 99-й — 27.30 руб.
- Выбросы по цене (> 27.30 руб.) составляют 1% выборки.

Набдюдаются следующие аномалии : отрицательное числа в  `Quantity` и крупное число позиций в `Quantity`.  

-----
-----

#// **Вот тут можно начать анализировть отрицательное количество и дальше думать что с ним делать**

## **Обработка  типов данных**

In [13]:
#осмотр данных

df.dtypes

,0
TransactionNo,object
Date,object
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,float64
Country,object


In [14]:
# преобразование данных
df['Date'] = pd.to_datetime(df['Date'])
df['CustomerNo'] = df['CustomerNo'].astype('Int64')

#проверка
df.dtypes

,0
TransactionNo,object
Date,datetime64[ns]
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,Int64
Country,object


### **Итог преобразования:**
- Колонка `Date` преобразована в `datetime` для работы с временными рядами
- Колонка `CustomerNo` преобразована  в `Int64` так как содержит пропуски (`NaN`)

Данные оптимизированы для анализа

---
----

## **Поиск и обработка пропусков**

In [15]:
# Пропуски: количество и доля
df.isnull().sum(), df.isnull().mean()

(TransactionNo     0
 Date              0
 ProductNo         0
 ProductName       0
 Price             0
 Quantity          0
 CustomerNo       55
 Country           0
 dtype: int64,
 TransactionNo    0.000000
 Date             0.000000
 ProductNo        0.000000
 ProductName      0.000000
 Price            0.000000
 Quantity         0.000000
 CustomerNo       0.000103
 Country          0.000000
 dtype: float64)

In [16]:
df[df['CustomerNo'].isnull()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,<NA>,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,<NA>,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,<NA>,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,<NA>,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,<NA>,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,<NA>,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,<NA>,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,<NA>,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,<NA>,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,<NA>,United Kingdom


In [17]:
# Замена NaN на специальное значение
df['CustomerNo'] = df['CustomerNo'].fillna(-1)

In [18]:

# строки, где были пропуски (
display(df[df['CustomerNo'] == -1].head(10))

# 2. Проверка
print(f"\nВсего замен: {(df['CustomerNo'] == -1).sum()}")

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,-1,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,-1,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,-1,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,-1,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,-1,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,-1,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,-1,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,-1,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,-1,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,-1,United Kingdom



Всего замен: 55


- Выбрана замена пустых значений для  сохранения данных



-----
-----

## **Поиск и дубликатов/неявных дубликатов**

In [19]:
df.duplicated().sum()

np.int64(5200)

In [20]:
# Процент дубликатов
print(f"Процент дубликатов: {(df.duplicated().sum() / len(df) * 100):.2f}%")

Процент дубликатов: 0.97%


In [21]:
# Первые 10 дубликатов
df[df.duplicated()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
985,581497,2019-12-09,21481,Fawn Blue Hot Water Bottle,7.24,1,17497,United Kingdom
1365,581538,2019-12-09,23275,Set Of 3 Hanging Owls Ollie Beak,6.19,1,14446,United Kingdom
1401,581538,2019-12-09,22992,Revolver Wooden Ruler,6.19,1,14446,United Kingdom
1406,581538,2019-12-09,22694,Wicker Star,6.19,1,14446,United Kingdom
1409,581538,2019-12-09,23343,Jumbo Bag Vintage Christmas,6.19,1,14446,United Kingdom
3255,581237,2019-12-08,23498,Classic Bicycle Clips,6.19,1,15443,United Kingdom
3259,581237,2019-12-08,23497,Classic Chrome Bicycle Bell,6.19,1,15443,United Kingdom
3261,581237,2019-12-08,23509,Mini Playing Cards Fun Fair,6.19,2,15443,United Kingdom
3284,581237,2019-12-08,23378,Pack Of 12 50'S Christmas Tissues,6.19,1,15443,United Kingdom
3593,581253,2019-12-08,23427,Stool Home Sweet Home,6.19,1,16891,United Kingdom


In [22]:

# Показать реальные дубликаты (первые 10)
df[df.duplicated(keep=False)].sort_values('TransactionNo').head(20)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
533720,536409,2018-12-01,22111,Scottie Dog Hot Water Bottle,15.32,1,17908,United Kingdom
533724,536409,2018-12-01,22866,Hand Warmer Scotty Dog Design,12.40,1,17908,United Kingdom
533729,536409,2018-12-01,21866,Union Jack Flag Luggage Tag,11.53,1,17908,United Kingdom
533752,536409,2018-12-01,21866,Union Jack Flag Luggage Tag,11.53,1,17908,United Kingdom
533756,536409,2018-12-01,22900,Set 2 Tea Towels I Love London,13.27,1,17908,United Kingdom
533761,536409,2018-12-01,22866,Hand Warmer Scotty Dog Design,12.40,1,17908,United Kingdom
533769,536409,2018-12-01,22900,Set 2 Tea Towels I Love London,13.27,1,17908,United Kingdom
533771,536409,2018-12-01,22111,Scottie Dog Hot Water Bottle,15.32,1,17908,United Kingdom
533832,536412,2018-12-01,22902,Tote Bag I Love London,12.40,7,17920,United Kingdom
533833,536412,2018-12-01,21448,12 Daisy Pegs In Wood Box,11.94,2,17920,United Kingdom


# // краткий коммент

----
----
----

#**Шаг 2. Предварительный анализ и создание новых признаков**

- Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
- Создание признаков: месяц, день недели, год из даты.
- Создание признака: общая сумма заказа (`Price` и `Quantity`).
- Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

# // **Создать колонку где будут лишь положительное колличество цен**

# **Шаг 3. Исследовательский анализ данных (EDA)**

*   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
*   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
*   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
*   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
*   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
*   Анализ ценовых сегментов: зависимость количества продаж от цены.

In [23]:
#топ-10 дорогих товаров

top_price = df.groupby('ProductName')['Price'].max().sort_values(ascending=False).head(10)
print(top_price)

ProductName
Vintage Red Kitchen Cabinet          660.62
Vintage Blue Kitchen Cabinet         660.62
Love Seat Antique White Metal        210.12
Rustic Seventeen Drawer Sideboard    179.37
Regency Mirror With Shutters         179.37
Chest Natural Wood 20 Drawers        138.38
Hall Cabinet With 3 Drawers          112.75
Sweetheart Cream Steel Table Rect    112.71
Cinderella Chandelier                112.71
Ivory Embroidered Quilt               94.81
Name: Price, dtype: float64


# **Шаг 4. Ключевые выводы и рекомендации**
*   Формирование основных инсайтов по каждому блоку анализа.
*   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.